README.md
# ⚡️ Supercharged

[![Pub](https://img.shields.io/pub/v/supercharged.svg)](https://pub.dartlang.org/packages/supercharged)
[![Tests](https://github.com/felixblaschke/supercharged/workflows/Tests/badge.svg)](https://github.com/felixblaschke/supercharged/actions?query=workflow%3ATests)

**Supercharged** brings **all the comfort features** from languages like Kotlin to **all Flutter developers**.

- 💪 **fully tested**
- 📝 **well documented**
- 💼 **enterprise-ready**

## ⛏️ Getting started

Add the dependency `supercharged: ^1.4.0` to your project and start using **Supercharged** everywhere:
```dart
import 'package:supercharged/supercharged.dart';
```

For more details have a look at the [install tab](https://pub.dev/packages/supercharged#-installing-tab-).


## 🌞 Highlights

Transform any `String` to **[colors](https://pub.dev/documentation/supercharged/latest/supercharged/String_/toColor.html)**
```dart
"#ff00ff".toColor(); // painless hex to color
"red".toColor(); // supports all web color names
```

Helpful [substring](https://pub.dev/documentation/supercharged/latest/supercharged/String_/allAfter.html) functions:
```dart
"mode:production".allAfter(":"); // "production"
"<html>".allBetween("<", ">"); // "html"`
"flutter is cool".allBefore(" is"); // "flutter"
```

[Parse](https://pub.dev/documentation/supercharged/latest/supercharged/String_/toDouble.html) user input:
```dart
"2.1".toDouble(); // 2.1
"42".toInt(); // 
```

Effortless [aggregation](https://pub.dev/documentation/supercharged/latest/supercharged/IterableOfInt_.html) for `Iterable<int>` and `Iterable<double>`:
```dart
[1, 2, 3].sum(); // 6
[1, 2, 3].average(); // 2
[1, 2, 3].min(); // 1
[1, 2, 3].max(); // 3
```

Advanced [aggregation](https://pub.dev/documentation/supercharged/latest/supercharged/Iterable_/sumBy.html) for any `Iterable`:
```dart
var persons = [Person(age: 20), Person(age: 30), Person(age: 40)];
persons.sumBy((p) => p.age); // 90
persons.averageBy((p) => p.age); // 30
persons.count((p) => p.age < 35); // 2
persons.minBy((a,b) => a.age.compareTo(b.age)); // Person(age: 20)
persons.maxBy((a,b) => a.age.compareTo(b.age)); // Person(age: 40)
```

Safely [access](https://pub.dev/documentation/supercharged/latest/supercharged/Iterable_/firstOrNull.html) `Iterable`:
```dart
[].firstOrNull(); // return null instead
[].lastOrNull();  // of throwing error

[].firstOrElse(() => Placeholder()); // Produce default values
[].lastOrElse(() => Placeholder());  // on the fly

[1, 2, 3].elementAtOrNull(4); // Never go out of bounds
[1, 2, 3].elementAtOrElse(4, () => 0);
```

[Group up data](https://pub.dev/documentation/supercharged/latest/supercharged/Iterable_/groupBy.html) to match view:
```dart
var persons = [
    Person(name: "John", age: 21),
    Person(name: "Carl", age: 18),
    Person(name: "Peter", age: 56),
    Person(name: "Sarah", age: 61)
];

persons.groupBy(
    (p) => p.age < 40 ? "young" : "old",
    valueTransform: (p) => p.name
); // {"young": ["John", "Carl"], "old": ["Peter", "Sarah"]}
```

[Chunking](https://pub.dev/documentation/supercharged/latest/supercharged/Iterable_/chunked.html) for easy pagination:
```dart
["a", "b", "c", "d", "e"].chunked(3); // [ ["a", "b", "c"], ["d", "e"] ]
["a", "b", "c"].chunked(2, fill: () => ""); // [ ["a", "b"], ["c", ""] ]
```

More natural [durations](https://pub.dev/documentation/supercharged/latest/supercharged/Int_/minutes.html) and dates:
```dart
var duration = 5.minutes + 30.seconds;
duration += 0.5.hours;

var future = 2.days.fromNow() + 20.minutes;
var past = 15.minutes.ago();

var allDaysOf2020 = DateTime(2020).until(DateTime(2021));
var allFullHoursOfFirstJanuary2020 =
  DateTime(2020, 1, 1).until(DateTime(2020, 1, 2), by: 1.hours);
```

Simplified [data sorting](https://pub.dev/documentation/supercharged/latest/supercharged/Iterable_/sortedByNum.html):
```dart
 persons = [
    Person(name: "John", age: 21),
    Person(name: "Carl", age: 18),
    Person(name: "Peter", age: 56),
    Person(name: "Sarah", age: 61)
  ];

  persons.sortedByNum((p) => p.age); // list sorted by age
  persons.sortedByString((p) => p.name); // list sorted by name
  persons  // sorted with comparator
      .sortedBy((a, b) => a.name.compareTo(b.name));

  persons // sorting is fully integrated into processing chain
      .filter((p) => p.name.length < 5)
      .sortedByNum((p) => p.age)
      .map((p) => p.name)
      .toList();
```

Solid tween shortcuts for animations:
```dart
100.0.tweenTo(200.0); // Tween(begin: 100.0, end: 200.0)
Colors.red.tweenTo(Colors.blue); // ColorTween(...)
0.0.tweenTo(100.0).curved(Curves.easeInOut); // Apply accelleration
```

[Replace](https://pub.dev/documentation/supercharged/latest/supercharged/Int_/rangeTo.html) your classic for-loop:
```dart
0.rangeTo(5); // [0, 1, 2, 3, 4, 5]
3.rangeTo(1); // [3, 2, 1]

var list = ["dog", "cat", "mouse"];
0.until(list.length); // [0, 1, 2]

list.forEachIndexed((index, value) {
    // index: 0, value: "dog"
    // index: 1, value: "cat"
    // index: 2, value: "mouse"
});
```

## 📑 API documentation

You can discover all **68 features** in the [API documentation](https://pub.dev/documentation/supercharged/latest/supercharged/supercharged-library.html).
Each feature is **well documented** with an **example**.

Feel free to quickly **jump** into a topic:

- [double](https://pub.dev/documentation/supercharged/latest/supercharged/Double_.html)
- [int](https://pub.dev/documentation/supercharged/latest/supercharged/Int_.html)
- [Iterable / List / Set](https://pub.dev/documentation/supercharged/latest/supercharged/Iterable_.html)
- [Iterable\<double\> / List\<double\> / Set\<double\>](https://pub.dev/documentation/supercharged/latest/supercharged/IterableOfDouble_.html)
- [Iterable\<int\> / List\<int\> / Set\<int\>](https://pub.dev/documentation/supercharged/latest/supercharged/IterableOfInt_.html)
- [Iterable\<MapEntry\> / List\<MapEntry\> / Set\<MapEntry\>](https://pub.dev/documentation/supercharged/latest/supercharged/IterableOfMapEntry_.html)
- [List](https://pub.dev/documentation/supercharged/latest/supercharged/List_.html)
- [Map](https://pub.dev/documentation/supercharged/latest/supercharged/Map_.html)
- [String](https://pub.dev/documentation/supercharged/latest/supercharged/String_.html)
- [Duration](https://pub.dev/documentation/supercharged/latest/supercharged/Duration_.html)
- [DateTime](https://pub.dev/documentation/supercharged/latest/supercharged/DateTime_.html)


## 📈 Improve

Supercharged will **improve** in future updates. Help me by reporting bugs, **submit new ideas** for features or anything else that you want to share.

- Just [write an issue](https://github.com/felixblaschke/supercharged/issues) on GitHub. ✏️
- And don't forget to hit the **like button** for this package ✌️


CHANGELOG.md
## 1.7.0

- **New:** DateTime `isBetween` method
- **Changed:** double `between` renamed to `isBetween`
- **Changed:** int `between` renamed to `isBetween`

## 1.6.0

- **New:** double `between` method
- **New:** int `between` method

## 1.5.1
- **Fixed:** Unit test for DateTime `duration` method

## 1.5.0

- **New:** DateTime `duration` method

## 1.4.0

- **New:** Animatable `animatedBy` method
- **New:** Animatable `curved` method

## 1.3.0

- **New:** int `tweenTo` method
- **New:** double `tweenTo` method
- **New:** Color `tweenTo` method
- **New:** Duration `/` operator

## 1.2.1

- **Fixed:** typo in API documentation

## 1.2.0

- **New:** DateTime `+` operator
- **New:** DateTime `-` operator
- **New:** DateTime `rangeTo` method
- **New:** DateTime `until` method
- **New:** Duration `fromNow` method
- **New:** Duration `ago` method
- **New:** Iterable `lastIndex` getter
- **New:** Iterable `withoutFirst` method
- **New:** Iterable `withoutLast` method
- **Improved:** Added `null` check for int `until`
- **Improved:** Added `null` check for int `times`
- **Improved:** Removed implicit casts and improved usage of strong types

## 1.1.0

- **New:** Iterable `sortedBy` method 
- **New:** Iterable `sortedByNum` method 
- **New:** Iterable `sortedByString` method 
- **New:** Iterable `chunked` method has additional parameter `fill`
- **Improved:** More efficient implementation for `minBy` and `maxBy`
- **Fixed:** typo in API documentation


## 1.0.0

* Initial release

example/main.dart
import 'package:flutter/material.dart';
import 'package:supercharged/supercharged.dart';

main() {
  // Transform any String to colors
  "#ff00ff".toColor(); // painless hex to color
  "red".toColor(); // supports all web color names

  // Helpful substring functions:
  "mode:production".allAfter(":"); // "production"
  "<html>".allBetween("<", ">"); // "html"`
  "flutter is cool".allBefore(" is"); // "flutter"

  // Parse user input:
  "2.1".toDouble(); // 2.1
  "42".toInt(); //

  // Effortless aggregation for Iterable<int> and Iterable<double>:
  [1, 2, 3].sum(); // 6
  [1, 2, 3].average(); // 2
  [1, 2, 3].min(); // 1
  [1, 2, 3].max(); // 3

  // Advanced aggregation for any Iterable:
  var persons = [Person(age: 20), Person(age: 30), Person(age: 40)];
  persons.sumBy((p) => p.age); // 90
  persons.averageBy((p) => p.age); // 30
  persons.count((p) => p.age < 35); // 2
  persons.minBy((a, b) => a.age.compareTo(b.age)); // Person(age: 20)
  persons.maxBy((a, b) => a.age.compareTo(b.age)); // Person(age: 40)

  // Safely access Iterable:
  [].firstOrNull(); // return null instead
  [].lastOrNull(); // of throwing error

  [].firstOrElse(() => Placeholder()); // Produce default values
  [].lastOrElse(() => Placeholder()); // on the fly

  [1, 2, 3].elementAtOrNull(4); // Never go out of bounds
  [1, 2, 3].elementAtOrElse(4, () => 0);

  // Group up data to match view:
  persons = [
    Person(name: "John", age: 21),
    Person(name: "Carl", age: 18),
    Person(name: "Peter", age: 56),
    Person(name: "Sarah", age: 61)
  ];

  persons.groupBy((p) => p.age < 40 ? "young" : "old",
      valueTransform: (p) =>
          p.name); // {"young": ["John", "Carl"], "old": ["Peter", "Sarah"]}

  // Chunking for easy pagination:
  ["a", "b", "c", "d", "e"].chunked(3); // [ ["a", "b", "c"], ["d", "e"] ]
  ["a", "b", "c"].chunked(2, fill: () => ""); // [ ["a", "b"], ["c", ""] ]

  // More natural durations and dates:
  var duration = 5.minutes + 30.seconds;
  duration += 0.5.hours;

  var future = 2.days.fromNow() + 20.minutes;
  var past = 15.minutes.ago();

  var allDaysOf2020 = DateTime(2020).until(DateTime(2021));
  var allFullHoursOfFirstJanuary2020 =
      DateTime(2020, 1, 1).until(DateTime(2020, 1, 2), by: 1.hours);

  // Simplified data sorting:
  persons = [
    Person(name: "John", age: 21),
    Person(name: "Carl", age: 18),
    Person(name: "Peter", age: 56),
    Person(name: "Sarah", age: 61)
  ];

  persons.sortedByNum((p) => p.age); // list sorted by age
  persons.sortedByString((p) => p.name); // list sorted by name
  persons
      .sortedBy((a, b) => a.name.compareTo(b.name)); // sorted with comparator

  persons // sorting is fully integrated into processing chain
      .filter((p) => p.name.length < 5)
      .sortedByNum((p) => p.age)
      .map((p) => p.name)
      .toList();

  // Solid tween shortcuts for animations
  100.0.tweenTo(200.0); // Tween(begin: 100.0, end: 200.0)
  Colors.red.tweenTo(Colors.blue); // ColorTween(...)

  // Replace your classic for loop:
  0.rangeTo(5); // [0, 1, 2, 3, 4, 5]
  3.rangeTo(1); // [3, 2, 1]

  var list = ["dog", "cat", "mouse"];
  0.until(list.length); // [0, 1, 2]

  list.forEachIndexed((index, value) {
    // index: 0, value: "dog"
    // index: 1, value: "cat"
    // index: 2, value: "mouse"
  });
}

class Person {
  final String name;
  final int age;

  Person({this.age, this.name = ""});
}

class Placeholder {}
